### PR TITLE
Beginning of neural ode in deepchem

### PR DESCRIPTION
**Pull request by shivance**
# Pull Request Template

## Description
This is the first PR for my contribution towards Convolutional Neural ODE model in deepchem <a href="https://github.com/deepchem/deepchem/issues/2680"> wishlist </a>

According to [this](https://forum.deepchem.io/t/towards-a-deep-science-suite/710) post by @rbharath on Deepchem’s Forum, “Growing support for Neural ODEs and CFD is on top priority list for Deepchem ''. The Convolutional ODE model to be contributed through this project in the original paper was used to model turbulent fluid flow by using Neural ODE

List any dependencies that are required for this change. -->
[torchdiffeq](https://github.com/rtqichen/torchdiffeq)

### CODE REVIEW

**Comment by shivance** [comment]
This is initial PR towards [this](https://docs.google.com/document/d/18u1_tAoDHFHu5-8yf6OkmGZIIjyFWtyYcjWMwXvjO9I/edit?usp=sharing) proposal. Need to add unit tests. Filtering of code is also pending. Will update soon ;)

**Comment by shivance** [comment]
@rbharath I need some help with writing unit tests. PS.: I went through model testing guide of deepchem.

**Comment by shivance** [comment]
@rbharath I've updated this PR to contain Encoder layer and it's unit test, it's ready for review

File: deepchem/models/tests/test_layers.py [hunk 1]
```diff
@@ -706,3 +704,17 @@ def test_mat_generator():
...
+
+
+@pytest.mark.torch
+def test_ConvODEEncoderLayer():
+  from deepchem.models.torch_models.layers import ConvODEEncoderLayer
+
+  _encoder = ConvODEEncoderLayer()
+  x = torch.rand(1,3,224,224)
+  y = _encoder(x)
+  assert list(y.shape) == [1,32,55,55], "Output Shape doesn't match expected shape"
```
Comment (rbharath) [inline deepchem/models/tests/test_layers.py:716]: assert y.shape == (1, 32, 55, 55), "Output Shape..."

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -851,3 +852,49 @@ def __repr__(self) -> str:
...
+    dim: int
+      Output dimension for the Encoder
+    """
+    super(ConvODEEncoderLayer, self).__init__()
+
+    self._encoder = None
+    if isinstance(encoder, nn.Module):
+      self._encoder = encoder
+
+    self.conv1 = nn.Conv2d(3, 16, 3, 2)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:878]: These should be options not hard coded

Comment (rbharath) [inline deepchem/models/torch_models/layers.py:878]: We should be able to control the number of convolutional layers (say 3 or 1 layer)

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -851,3 +852,49 @@ def __repr__(self) -> str:
...
+
+    encoder: nn.Module
+      User Specified Encoder Block, can be passed while initializing
+    dim: int
+      Output dimension for the Encoder
+    """
+    super(ConvODEEncoderLayer, self).__init__()
+
+    self._encoder = None
+    if isinstance(encoder, nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:875]: No need to pass in encoder. User can just encode it themselves before calling your layer

File: deepchem/models/tests/test_layers.py [hunk 4]
```diff
@@ -706,3 +704,17 @@ def test_mat_generator():
...
+def test_ConvEncoderLayer():
+  """Test invoking ConvEncoderLayer"""
+  torch.manual_seed(0)
+
+  from deepchem.models.torch_models.layers import ConvEncoderLayer
+
+  encoder = ConvEncoderLayer(layer_filters=[3,16,8,32,16],dims=2,kernel_size=4)
+  x = torch.ones(1, 3, 224, 224)
+  y = encoder(x)
+  assert list(y.shape) == [16], "Output Shape doesn't match expected shape"
```
Comment (arunppsg) [inline deepchem/models/tests/test_layers.py:719]: assert y.shape == (16, ) something like this should be fine

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -851,3 +856,145 @@ def __repr__(self) -> str:
...
+  - A global pooling layer (either max pool or average pool)
+
+  Examples
+  --------
+  >>> from deepchem.models.torch_models.layers import ConvEncoderLayer
+  >>> import deepchem as dc
+  >>> encoder = ConvEncoderLayer(layer_filters=[3,16,8,32,16], conv_dim=3, kernel_size=4)
+  >>> x = torch.ones(3,224,224,224)
+  >>> y = encoder(x)
+  >>> y = tensor([1.2733, 0.0000, 0.1453, 0.8308, 0.0000, -0.0000, 0.7857, -0.0000, 1.2164,
```
Comment (arunppsg) [inline deepchem/models/torch_models/layers.py:876]: >>> len(y)
16
>>> y.shape
1, 16

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -853,6 +859,140 @@ def __repr__(self) -> str:
     )
 
 
+class ConvEncoderLayer(nn.Module):
+  """A 1,2, or  dimensional convolutional network for Convolutional Neural ODE Model
```
Comment (arunppsg) [inline deepchem/models/torch_models/layers.py:863]: A 1,2 or 3

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -853,6 +859,140 @@ def __repr__(self) -> str:
...
+  >>> y = encoder(x)
+  >>> y.shape
+  torch.Size([5,16])
+
+  """
+
+  def __init__(self,
+               conv_dim: int,
+               layer_filters: List[int],
+               kernel_size: Optional[int or List[int]] = 5,
```
Comment (arunppsg) [inline deepchem/models/torch_models/layers.py:883]: Optional[Union[int, List[int]]]

**Comment by shivance** [comment]
@rbharath Deepchem Core Build CI is failing :
```
Run mypy -p deepchem
deepchem/models/torch_models/layers.py:886: error: Incompatible default for argument "dropouts" (default has type "float", argument has type "Union[int, List[int], None]")
deepchem/models/torch_models/layers.py:887: error: Incompatible default for argument "activation_fns" (default has type "Type[ReLU]", argument has type "Union[Module, List[Module], None]")
deepchem/models/torch_models/layers.py:931: error: List item 0 has incompatible type "Optional[int]"; expected "int"
deepchem/models/torch_models/layers.py:934: error: List item 0 has incompatible type "Optional[int]"; expected "int"
deepchem/models/torch_models/layers.py:937: error: List item 0 has incompatible type "Optional[int]"; expected "int"
deepchem/models/torch_models/layers.py:940: error: List item 0 has incompatible type "Optional[Module]"; expected Module
deepchem/models/torch_models/layers.py:948: error: Incompatible types in assignment (expression has type "Union[Type[MaxPool1d], Type[MaxPool2d], Type[MaxPool3d]]", variable has type "Union[Type[AvgPool1d], Type[AvgPool2d], Type[AvgPool3d]]")
deepchem/models/torch_models/layers.py:965: error: Argument "padding" to "Conv1d" has incompatible type "Optional[str]"; expected "Union[str, Union[int, Tuple[int]]]"
deepchem/models/torch_models/layers.py:965: error: Argument "padding" to "Conv2d" has incompatible type "Optional[str]"; expected "Union[str, Union[int, Tuple[int, int]]]"
deepchem/models/torch_models/layers.py:965: error: Argument "padding" to "Conv3d" has incompatible type "Optional[str]"; expected "Union[str, Union[int, Tuple[int, int, int]]]"
Found 10 errors in 1 file (checked 357 source files)
Error: Process completed with exit code 1.
```

What should be Type for "padding"?

**Comment by shivance** [comment]
@rbharath suggested to port CNN to PyTorch and use that as a ConvEncoder Layer.
The implementation pretty much remain the same. Will be working with @peastman on porting.
New PR for Torch-CNN : https://github.com/deepchem/deepchem/pull/2950
